RedJubjub

str4d opened the pull request.

str4d commented:
Force-pushed the second commit to fix bugs and refactor slightly.

gtank commented on src/redjubjub.rs:
```diff
@@ -0,0 +1,133 @@
+            .expect("Jubjub points should serialize to 32 bytes");
+        let c = h_star::<E>(&rbar[..], msg);
+
+        // S . G = R + c . vk
```
Do we want to follow the RFC for EdDSA? In that case we need to check for the canonical signature representation, i.e. that 0 <= S < order of the curve.

gtank replied:
This is in the spec definition of RedDSA.Verify already, but typo'd - we should ensure S < $\ParamJ{r}

gtank replied:
Additional comment because I forget to start a review

str4d commented:
Addressed @gtank's comments. In particular, I added a comment that the other signature validity checks will occur during signature de-serialization (once that is implemented).

gtank commented on src/redjubjub.rs:
```diff
@@ -80,6 +80,8 @@ impl<E: JubjubEngine> PublicKey<E> {
         let c = h_star::<E>(&rbar[..], msg);
 
         // S . G = R + c . vk
+        // The other signature validation checks are performed
```
Please specify what you mean. 'Specifically, the field type enforces the range of the scalars' or whatever is actually going to be true. If I'm an auditor diffing this against the RFC, I'll be looking for specific mention of the preconditions. If I don't see it, I might waste engagement time chasing a malleability red herring.

gtank commented:
LGTM

daira commented on src/redjubjub.rs:
```diff
@@ -0,0 +1,133 @@
+}
+
+pub struct Signature<E: JubjubEngine> {
+    r: Point<E, Unknown>,
```
It does not (or even large order). The spec explicitly says not. It doesn't in EdDSA either.

The verification equation is [S] G = R + [H*(R || M)] vk, where G is the base and vk is the verifying key. For the binding signature, vk is calculated in a way that ensures it is in the prime order subgroup (it might be the zero point, theoretically, but that would break the Pedersen hash). For the spend authorization signature we don't currently have any checks other than that vk is a valid point.

daira commented on src/redjubjub.rs:
```diff
@@ -0,0 +1,137 @@
+        let mut rbar = [0u8; 32];
+        sig.r
+            .write(&mut rbar[..])
+            .expect("Jubjub points should serialize to 32 bytes");
```
If I'm reading correctly, we're decoding a point representation (outside this class), reencoding it here, and relying on the fact that other code enforced a canonical representation. That seems ugly and error-prone.

str4d replied:
This is an artifact of @ebfull's desire to represent `Signature` in-memory as the unserialized values. We could alternatively represent it as the serialized string, and unserialize it every time we verify the signature. Or we could store both `Rbar` and `R` in the signature, which I don't like because it feels like it could get out-of-sync.

daira replied:
It was an intentional part of the design (inherited from EdDSA) that you hash the representation of R exactly as given in the signature. That implies that any change to the representation R̲ will result in a completely different hash, and so it's not possible to malleate the signature by changing R̲. (It still might be by changing S̲, but that's a simpler encoding.)

I realize it is equivalent if the representation of R is ensured to be canonical, but then you have two places to look in the code to check that the signature is being verified correctly. As specified, a RedJubjub signature is a byte sequence, *not* a (point, integer) pair, and so I recommend that it be implemented that way for ease of auditing.
